### PR TITLE
RCHAIN-3503: Implement basic lock

### PR DESCRIPTION
**Pull request by JosDenmark**
## Overview
This is the basic lock needed to write the readers-writer lock (upcoming PR).

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-3503

### CODE REVIEW

**Comment by JosDenmark** [comment]
bors try

**Comment by JosDenmark** [comment]
@KentShikama It's not that different. It's basically a purely functional wrapper around a SyncVar that holds `Unit`.

**Comment by lukasz-golebiewski** [comment]
Isn't this code functionally equivalent to what a `Semaphore` with a single permit offers?

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/error/MLockSpec.scala [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+import scala.concurrent.duration._
+
+class MLockSpec extends FlatSpec with Matchers {
+
+  import monix.execution.Scheduler.Implicits.global
+
+  private val MAX_DURATION = 10.seconds
+
+  def withFixture[A](f: MLock[Task] => Task[A]): A =
```
Comment (lukasz-golebiewski) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/error/MLockSpec.scala]: `in fixture` reads better than `in withFixture`

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/error/MLockSpec.scala [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+      .map(_ should be(Left(LockReleaseException("attempted to release released lock"))))
+  )
+
+  "release" should "unblock lock acquisitions in FIFO order" in withFixture { lock =>
+    for {
+      ref <- Ref.of[Task, Int](0)
+      _   <- lock.acquire
+      result <- List(
+                 lock.acquire >> ref.getAndSet(1),
+                 Task.sleep(0.1.seconds) >> lock.acquire >> ref.getAndSet(2),
```
Comment (lukasz-golebiewski) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/error/MLockSpec.scala]: Using `sleep` can lead to potential race conditions and test failures in rare cases. Consider using this approach instead: https://monix.io/docs/3x/execution/scheduler.html#injecting-time-and-tests

**Comment by JosDenmark** [comment]
@lukasz-golebiewski What would be the analogue to `waitForLock`?

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/error/MLockSpec.scala [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    } yield assert(result.head == 0 && result(1) == 0 && result(2) == 1)
+  }
+
+  "release" should "unblock all waiting threads simultaneously" in withFixture { lock =>
+    for {
+      ref   <- Ref.of[Task, Int](0)
+      _     <- lock.acquire
+      count = 999
+      _ <- List(
+            (0 to count).toList.parTraverse_(_ => lock.waitForLock >> ref.update(_ + 1)),
```
Comment (lukasz-golebiewski) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/error/MLockSpec.scala]: I'll answer to your comment https://github.com/rchain/rchain/pull/2504#issuecomment-500872011 here as this is the only usage of `lock.waitForLock`. If you remove `lock.waitForLock >>` from the test above it still passes. This is due to the nature of `ref.update` as stated in the docs:
```/**
   * Modifies the current value using the supplied update function. If another modification
   * occurs between the time the current value is read and subsequently updated, the modification
   * is retried using the new value. Hence, `f` may be invoked multiple times.
   *
   * Satisfies:
   *   `r.update(_ => a) == r.set(a)`
   */
  def update(f: A => A): F[Unit]
```

Maybe the functionality of `waitForLock` isn't absolutely necessary?

Comment (JosDenmark) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/error/MLockSpec.scala]: Its functionality is required in a following PR:

```
final class RWLock[F[_]](stateRef: MVar[F, State], readLock: MLock[F], writeLock: MLock[F])(
    implicit bracketF: Bracket[F, Throwable]
) {

  /**
    * Acquires read lock.
    *
    * Blocks (asynchronously) if another thread has write lock.
    *
    */
  def acquireRead: F[Unit] =
    stateRef.take >>= {
      case Free        => readLock.acquire >> stateRef.put(Read(1))
      case Write       => stateRef.put(Write) >> writeLock.waitForLock >> acquireRead
      case Read(count) => stateRef.put(Read(count + 1))
    }
...
}

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/error/MLock.scala [hunk 4]
```diff
@@ -0,0 +1,41 @@
+package coop.rchain.rholang.interpreter.error
+
+import cats.effect.{Bracket, Concurrent}
+import cats.implicits._
+import cats.effect.concurrent.MVar
+
+final class MLock[F[_]](lock: MVar[F, Unit])(implicit bracketF: Bracket[F, Throwable]) {
+
+  def acquire: F[Unit] = lock.take
```
Comment (odeac) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/error/MLock.scala]: Do we really want this low level API? What happens is we only expose a resource like API?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/error/MLock.scala [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    */
+  def release: F[Unit] =
+    lock
+      .tryPut(())
+      .ifM(
+        ().pure[F],
+        LockReleaseException("attempted to release released lock").raiseError[F, Unit]
+      )
+
+  def withLock[A](fa: F[A]): F[A] = bracketF.bracket[Unit, A](acquire)(_ => fa)(_ => release)
```
Comment (odeac) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/error/MLock.scala]: Using cats resource would be better for composability

**Pull request closed without merging**